### DISCUSSION TITLE
How does bpftrace deal with system-wide headers?

### DISCUSSION THREAD

**Discussion by nela** [bpftrace/bpftrace#2452] [Q&A] [answered]
Hello,

I am currently experimenting with different BPF tools and frontends. For this purpose I've been experimenting with uprobes, attached to an nginx process. I've been able to define a simple uprobe in bpftrace like:
```
#include "/usr/include/stdint.h"
#include "ngx_http.h"

uprobe:/usr/sbin/nginx:ngx_http_finalize_request
{
    $req = (struct ngx_http_request_s *)arg0;
    printf("Req line: %s\n", str($req->request_line.data));
}
```

Now, I am trying to do similar tracing directly with libbpf, but am getting strange errors  when including nginx header files, the build does not register my 64-bit architecture and wants to include `gnu/stubs-32.h`.

```
❯ make
  BPF      .output/nginx.bpf.o
In file included from nginx.bpf.c:4:
In file included from /home/nela/master/nginx-1.18.0/src/http/ngx_http.h:12:
In file included from /home/nela/master/nginx-1.18.0/src/core/ngx_config.h:26:
In file included from /home/nela/master/nginx-1.18.0/src/os/unix/ngx_linux_config.h:18:
In file included from /usr/include/x86_64-linux-gnu/sys/types.h:25:
In file included from /usr/include/features.h:510:
/usr/include/x86_64-linux-gnu/gnu/stubs.h:7:12: fatal error: 'gnu/stubs-32.h' file not found
 # include <gnu/stubs-32.h>
           ^~~~~~~~~~~~~~~~
1 error generated.
make: *** [Makefile:106: .output/nginx.bpf.o] Error 1
```

I have [posted a discussion](https://github.com/libbpf/libbpf-bootstrap/discussions/133#discussioncomment-4365389)  (which includes the code in question) on libbpf repo, but didn't really figure out how to solve the issue. I am wondering how does bpftrace deal with system-wide headers? More specifically, these errors seem to stem from inclusion of `sys/types.h` and `features.h`. I am having troubles including this in both BCC and strict libbpf code. What kind of magic does bpftrace do to resolve type redefinitions, and architecture detection? And if it is something that I can include in my libbpf code to solve?

**Comment by viktormalik** [comment, marked as answer]
Hi, bpftrace doesn't do much magic concerning the headers. There are few headers which are built-in inside bpftrace (see [resources/](https://github.com/iovisor/bpftrace/tree/master/resources)) but `sys/types.h` is not one of them. Other than that, we simply call the Clang preprocessor, so I don't see what the problem could be.

**Reply by nela** [reply]
thanks for the reply. Yeah, the issue is a bit strange to me / so no idea how I would go about fixing it. The obvious solution is not to use a whole lot of system header with bpf, but bpftrace seems to handle it quite well
